substrate/frame/treasury/src/benchmarking.rs

```diff
@@ -104,12 +104,39 @@ fn create_spend_arguments<T: Config<I>, I: 'static>(
 	(asset_kind, 100u32.into(), beneficiary, beneficiary_lookup)
 }
 
+// Use this `spend` insertion to bypass the `SpendOrigin` check. This is helpful
+// for benchmarking `payout`, `check_status`, and `void_spend`, which require
+// a spend in storage even if `SpendOrigin` cannot provide a successful origin.
+fn force_add_spend<T: Config<I>, I: 'static>(
+	asset_kind: Box<T::AssetKind>,
+	amount: AssetBalanceOf<T, I>,
+	beneficiary: Box<BeneficiaryLookupOf<T, I>>,
+) -> Result<(), &'static str> {
```

Comment:
This is not correct. What you should do is to skip adding something to `Spend` when `T::SpendOrigin::try_successful_origin()` fails. 

So,
```
		if let Ok(origin) = T::SpendOrigin::try_successful_origin() {
		Treasury::<T, _>::spend(
			origin,
		force_add_spend::<T, _>(
			Box::new(asset_kind.clone()),
			amount,
			Box::new(beneficiary_lookup),
			None,
		)?;
}
```

is the logic I would expect for these benchmarks below.

Reply:
If we omit adding something to the spend, the benchmark will fail during execution with the 'InvalidIndex' error. While it is possible to bypass this error using a solution like [this](https://substrate.stackexchange.com/questions/8215/benchmark-failing-extrinsics), it appears error-prone.

As extrinsic benchmark cover worst case scenario, It seems not far-fetched that even with `SpendOrigin = frame_support::traits::NeverEnsureOrigin<u64>` we take this impossible case.

Reply:
I mean we should be able to use the same condition as above to check either on success or error?

CC @ggwpez

Reply:
I know that the `#[extrinsic_call]` cannot be included inside a conditional block.

If there is a way to perform a benchmark with a condition, I have never seen it.

Reply:
You can use `#[block]` and put the `if` inside:  
```rust
#[block]
{
    if condition {
        ...
    } else {
        ...
    }
}
```
Condition should be a `bool` already to not mess up the timing. Otherwise you can create a new benchmark just for the failing case and then use the maximum of the two as weight.

Reply:
Quite interesting, I will implement that. I see in the documentation that "the #[block] syntax is broad and will benchmark any code contained within the block the attribute is attached to".  So, we consider the if condition check negligible in the timing, right?

Reply:
Yea negligible, but since we measure worst case timing it is actually not harming us since it should only slow it down.

Reply:
I understand, but I have a conceptual issue with these benchmarks because the worst-case scenario for them (most general case) is when the extrinsic finds a spend and performs what it should do. In the (edge) case where `SpendOrigin = frame_support::traits::NeverEnsureOrigin<u64>;`, no spend can ever be added to the Storage. As a result, with this setting, these three extrinsics should never be used by clients, except in cases of error calls.

Complicating the benchmark with conditionals in a solution that slightly overestimate the weight versus charging the most general case where a spend is found in storage (also overestimating weight if this is an impossible case to obtain in production) tends to yield the same result.

I'm not sure if I'm explaining myself very clearly.

Reply:
> As a result, with this setting, these three extrinsics should never be used by clients, except in cases of error calls.

Yes, no one should use them. However, an attacker will may want to use them to spam your chain. This is the reason that it needs a proper weight, aka account for the read.

Reply:
The more general case (processing the Spend) as a weight that is superior to the edge case (one read then an error), thus posing no risk of spamming…

Reply:
> The worst-case scenario occurs when a valid spend is found and paid.

But this isn't the worst-case when you have `SpendOrigin = NeverEnsureOrigin<u64>;`. Then the worst-case is basically the best case as well and just means that there is only one read of the storage.

Reply:
> > The worst-case scenario occurs when a valid spend is found and paid.
> 
> But this isn't the worst-case when you have `SpendOrigin = NeverEnsureOrigin<u64>;`. Then the worst-case is basically the best case as well and just means that there is only one read of the storage.

From the extrinsic point of view, it is the worst path because there is no mention of `SpendOrigin` in its implementation. In all the benchmarks that I found in Substrate, the worst path is determined from the extrinsic perspective, rather than considering all the (sometimes complicated) interference of all the parameters that can play a role outside the extrinsic implementation (and that can be quite complicated to anticipate when writing the benchmark, see this error for example).

I think that the implementation you're asking is suboptimal by adding unnecessary complexity to the benchmark for something that doesn't pose any threat to the chain but add a slight overweight. However, you seem very committed to this specific approach, so I will move this pull request to draft status and I will implement it according to your preferences when I have some time.

Reply:
Not sure your arguments are correct here. I mean the problem here is that the **benchmark** is failing because it is written in an incorrect way. The benchmark is "abusing" `SpendOrigin` to setup the state and the benchmark is failing because `SpendOrigin` is set to `NeverEnsureOrigin`. Now if we fix the **benchmark**, we should do it properly that it generates the correct weights. The thing I'm proposing here is also not really that complicated and can be implemented in some minutes.

Reply:
Yes, I agree that the benchmark is failing because it was written incorrectly. In my opinion, this failure directly stems from not approaching it from the extrinsic point of view, as I explained before. 

No, it's not complicated. However, since **I'm paid to work on another Substrate chain, this fix isn't a priority in my work schedule** that is quite full at the moment. Feel free to complete it if you want it done rapidly (i.e. before the end of a week).